### ISSUE TITLE
Task 2: component offload via moe-store groups + AdaLN caching + FP8 (24 GB target)

### ISSUE THREAD

**Issue by drunkcoding** [EfficientMoE/Omni-Infinity#2]
Third bootstrap task from EfficientMoE/MoE-Infinity#222. All inputs are now validated and in place:

## Ready inputs
- **v2 store of real H3-Base**: `/mnt/raid0nvme0/leyang/h3-store` (dev box) — 140 groups / 3486 tensors / 129 GB; 50 per-block non-AdaLN groups + 50 AdaLN bundle groups + per-shard VAE/encoder groups; byte-exact verified (EfficientMoE/moe-store#9).
- **Golden latents** (`tests/fixtures/goldens/fl2va_goldens.pt`): bitwise-validated reference, 256p/120f/8-step/seed-0.
- **Baseline to beat**: diffusers `ComponentsManager` sequential CPU offload already runs the 144 GB set on one 96 GB GPU, ~30-50 s denoise at 256p/120f. Task 2's win must come from one-read group streaming, AdaLN branch caching (no 13B re-transfer per component swap), and FP8 to reach ≤24 GB.

## Scope (RFC v0.1 memory story)
1. `StoreComponentSource`: materialize pipeline components from moe-store groups (sequential one-read fetches) instead of safetensors files.
2. AdaLN branch cache: keep `transformer_blocks.N.adaln` bundles host-resident; materialize on demand during denoise.
3. FP8 weight path on moe-store quantized group support (invariant G4).

## QA gate (RFC)
On a single 24 GB GPU: FL2VA smoke with `--max-vram 22GiB` → exit 0; `torch.cuda.max_memory_allocated() < 22 GiB` asserted in-script; latents within `allclose(rtol=2e-2)` of the Task 1 goldens; wall-clock recorded.

Note for the harness: dev-box GPUs are 96 GB — the 24 GB envelope should be emulated via the max-vram budget assertion, with a true-24-GB-card run before release.

**Comment by drunkcoding** [comment]
**Increment 1 landed** (0430836, CI green): `StoreComponentSource` — one-read group fetches from the moe-store v2 store.

- `read_group()`: single contiguous read at `(offset, total_size)`, member slicing in memory (asserted by a monkeypatched read counter in CI)
- `load_component_state_dict()`: diffusers-namespace state dicts per component; `adaln_groups()` exposes the 50 host-cacheable bundles
- Validated against the real 129 GB store: component set, 50 AdaLN bundles, `audio_vae` value-equality vs source

**Finding for the FP8 work**: the converter's `CastPolicy` already casts fp32 source params (audio_vae/vae) to bf16 in the store — source comparisons must be cast-aware, and the FP8 path will extend this same cast seam.

**Next increments**: (2) wire `StoreComponentSource` into the runner via `pipeline.update_components` + golden re-validation; (3) AdaLN branch cache with per-step materialization; (4) FP8 groups + the 22 GiB `max_memory_allocated` gate.

**Comment by drunkcoding** [comment]
**Increment 2 (VAE scope) landed and validated**: `from_pretrained(store_dir=...)` builds `vae` + `audio_vae` from moe-store groups and injects them via `update_components()`; the golden parity gate **passes bitwise** with store-sourced components (fresh generation, 65 s). The identical seam extends to `transformer`/`text_encoder` (increment 2b) — then AdaLN cache (3) and FP8 + 22 GiB gate (4).

**Comment by drunkcoding** [comment]
**Increment 2b landed and validated — the meaningful one.**

The transformer (66 GB, the actual latent producer) now loads from the moe-store and reproduces the golden latents **bitwise** through an 8-step generation. Full state_dict + buffers match the reference transformer exactly.

Getting there surfaced and fixed a real bug (EfficientMoE/moe-store#11, merged):
- **Correction to increment 2**: the VAE-only parity pass was weak — VAEs run *after* the latent goldens, so they never affected the check. The transformer is the real test.
- **Root cause**: moe-store's pipeline converter applied a blanket compute-dtype cast, truncating H3's intentionally-fp32 modules (`proj_in`/`time_embedder`/`proj_out`/`rope`) to bf16. This desynced `rope.inv_freq` (a non-persistent buffer computed from those fp32 weights), diverging the latents. Fixed by dtype-preserving pipeline conversion + a loader that snapshots `_keep_in_fp32_modules` tensors before the compute-dtype cast.

**State of Task 2**: increments 1, 2, 2b done — the store is a verified bitwise drop-in for all latent-producing components. Remaining: (3) AdaLN branch cache (host-resident bundles, per-step materialization), (4) FP8 groups + the 22 GiB `max_memory_allocated` gate. FP8 layers on the now-correct dtype-preserving baseline via moe-store quantized groups (G4).

**Comment by drunkcoding** [comment]
**Increment 3 landed and validated** (f664956): host-resident AdaLN branch cache.

The 50 per-block AdaLN projections (`transformer_blocks.N.adaln_proj`, ~13B params / **26.02 GB bf16** — ~40% of the 66 GB transformer) are now held in host memory and materialized on the compute device on demand, instead of being GPU-resident.

- `HostResidentAdaLN` replaces each block's `adaln_proj`, holding the weight/bias as a **plain attribute** (invisible to `Module.to(device)`, so `pipeline.to(device)` leaves it on the host) and materializing per forward. Its forward is a byte-for-byte reproduction of `MiniMaxH3AdaLayerNormModulation` (silu at `temb`'s precision → cast to the projection dtype → linear → six-way split).
- `load_transformer_with_adaln_cache` swaps the 50 modules **before** the state-dict load and the compute-dtype cast, so the AdaLN branches never become GPU-movable `Parameter`s and the discarded fp32 init Linears are freed up front. Built real-on-CPU (not meta-init) to keep the computed non-persistent `rope.inv_freq` buffer intact (the inc-2b desync class).
- `read_adaln_cache` keys blocks by the index parsed from the member name — the converter leaves `group.layer_id == -1` for the dense H3 groups.

**Validation** (real 129 GB store, RTX PRO 6000): a full packed forward through the AdaLN-cached transformer reproduces the full-resident store transformer's **video and audio outputs bitwise**, while GPU-resident params drop **66.28 → 40.26 GB** (the 26 GB of branches are now host-resident). Combined with inc 2b (store transformer → goldens bitwise), that is golden parity for the cached path. `OMNI_H3_ADALN_CACHE=1` selects the cache in the parity gate.

**Finding for the FP8 work**: `AdaLNEntry` pins the compute dtype (bf16) independent of storage dtype and carries an optional `scale`, so an FP8 quantized bundle dequantizes inside `materialize()` with the forward pass unchanged — increment 4 is a materialize-only change on this seam.

**State of Task 2**: increments 1, 2, 2b, 3 done. Note the tradeoff — inc 3 is synchronous single-stream H2D (~26 GB/step), trading VRAM for bandwidth; stream overlap is a separable perf increment. Remaining: **(4) FP8 groups + the 22 GiB `max_memory_allocated` gate** — FP8 layers on this dtype-preserving baseline via moe-store quantized groups (G4) and also halves the AdaLN H2D traffic.

**Comment by drunkcoding** [comment]
**Increment 4 landed** (1156750): FP8 weight path (opt-in) + offline QA harness. **The FP8 memory lever works but cannot meet the RFC's `rtol=2e-2` latent gate** — evidence and the deferred path below.

## What landed
- `omni_infinity/fp8.py`: `ScaledFp8Linear` (per-output-channel float8 weight + scale, upcast per forward) and `apply_scaled_fp8_casting(skip_blocks=…)`; `read_adaln_cache(fp8=)` applies the same per-row scheme to the AdaLN bundles. diffusers' `enable_layerwise_casting` was tried and rejected (unscaled float8 → **17–35%** off on H3's ~1e-2 weights).
- `fl2va_smoke.py`: `--transformer-fp8`, `--fp8-skip-last-blocks`, `--max-vram`, `--goldens`; a **denoise-window** `max_memory_allocated` probe (reset at first transformer forward, captured at the last — so text-encode/VAE-decode don't inflate it) and `rms_rel`-vs-goldens reporting.
- Two fixes that make the **offline end-to-end smoke** runnable: load the Qwen3-VL processor from the checkpoint path (its spec resolves by repo id, which raises under HF offline mode), and enable `ComponentsManager` auto-offload **after** `update_components`. `memory_reserve_margin = (device total − budget)` emulates a small-VRAM envelope on the 96 GB dev box.

## Validated on the real store + GPU
- FP8 cuts the transformer's resident weights **40.26 → 20.19 GB**.
- **Accuracy fails `rtol=2e-2`.** H3 is **not FP8-native** (bf16 checkpoint, no quantization config, store dtypes = {fp32, bf16}), so post-hoc FP8 is lossy through the 50-block stack:

  | config | FP8 blocks | latent rms_rel vs bf16 goldens |
  |---|---|---|
  | full-transformer FP8 | 0–49 | **~25%** |
  | selective, keep last 24 bf16 | 0–25 | **~16%** |

  Per-block profiling shows why: cumulative divergence stays <2% through block ~25, then **explodes in the last ~20 blocks** (0.07 → 0.15 → 0.23 → 0.54). The bf16 late blocks don't *add* FP8 error but **amplify** the error the early FP8 blocks seed, so no selective split reaches 2e-2 at a useful memory point.
- **Kernel note**: `torch._scaled_mm` FP8 GEMM is available on these Blackwell GPUs (cap 12.0), but true FP8 GEMM quantizes **activations** too — adding error on top of the weight-only path that is already the most accurate variant. FP8 GEMM helps latency, not this accuracy gate.

`transformer_fp8` therefore **defaults OFF** — it is a documented memory/accuracy tradeoff, not a gate-passing path.

## State of Task 2
Increments 1, 2, 2b, 3 done and bitwise-validated. Increment 4 lands the FP8 infra + QA harness and **conclusively characterizes** the FP8 accuracy wall. The RFC's `rtol=2e-2` + `<=22 GiB` gate is **deferred**: the accuracy-preserving route to the 24 GB envelope is **bf16 block-streaming** (keeps inc-3 bitwise parity) + **budget-enforced offload** (the `memory_reserve_margin` seam is in place), not FP8. AdaLN-cache host FP8 remains available to halve host RAM + H2D independently.

**Comment by drunkcoding** [comment]
**Increment 7 landed** ([#6](https://github.com/EfficientMoE/Omni-Infinity/pull/6), CI green): the whole-pipeline `<22 GiB` RFC QA gate.

## What landed
- `fl2va_smoke.py` now accepts `--vram-window {denoise,full}`. `denoise` remains the default, preserving increment 4 behavior.
- `full` resets `torch.cuda.max_memory_allocated()` immediately before `runner.generate(...)`, reads it immediately after, and uses that full text-encode + denoise + VAE-decode peak for `--max-vram`.
- CPU-safe tests cover the CLI default and the reset → generate → peak-read ordering.
- The README records the measured result and keeps the physical 24 GB-card confirmation explicit.

## Real-store/GPU validation
GPU 1 was the freest shared RTX PRO 6000 Blackwell Server Edition card (50,046 MiB free before launch). The existing `memory_reserve_margin = device total - budget` path emulated the 22 GiB envelope on the 96 GB card.

```bash
CUDA_VISIBLE_DEVICES=1 HF_HOME=/mnt/raid0nvme0/leyang/hf-home HF_HUB_OFFLINE=1 TRANSFORMERS_OFFLINE=1 PYTHONPATH=/mnt/raid0nvme0/leyang/Omni-Infinity python examples/fl2va_smoke.py --prompt "a red ball bouncing" --seed 0 --steps 8 --resolution 256p --frames 120 --offload --checkpoint /mnt/raid0nvme0/leyang/hf-home/hub/models--MiniMaxAI--MiniMax-H3/snapshots/42ed227ee7df40d41602854ae760620d6eb651fe --store-dir /mnt/raid0nvme0/leyang/h3-store --store-components transformer,vae,audio_vae --adaln-host-cache --block-stream-blocks-per-group 1 --stream-text-encoder --vram-window full --max-vram 22GiB --goldens tests/fixtures/goldens/fl2va_goldens.pt
```

```text
generate wall-clock: 38.1s
latents vs goldens: global rms_rel=0.0000 elementwise_allclose(rtol=2e-2,atol=2e-2)=True
full pipeline-window max_memory_allocated: 9.96 GiB (budget 22.00 GiB) ok=True
```

The emulated whole-pipeline gate **passes**: **9.96 GiB**, bitwise-equivalent latent result (`rms_rel=0.0000`, elementwise allclose true), **38.1 s**. No FP8 accuracy tradeoff is involved; this composes store-sourced components, host-resident AdaLN, bf16 transformer block-streaming, and bf16 Qwen3-VL leaf streaming.

## State of Task 2
Increments 1–7 are landed. The RFC's in-script full-window memory/parity gate is complete under the budget-enforced 96 GB dev-box emulation. The only remaining pre-release validation is confirmation on a **physical 24 GB card**; this result is not represented as that hardware run.
